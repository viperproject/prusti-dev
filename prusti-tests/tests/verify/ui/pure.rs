// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Phide_uuids=true
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// normalize-stdout-test: "\[[a-z0-9]{4}\]::" -> "[$(CRATE_ID)]::"

use prusti_contracts::*;

#[pure]
fn identity(x: u32) -> u32 { x }

fn test_identity1() {
    assert!(5 == identity(5));
}

#[ensures(6 == identity(6))]
fn test_identity2() {
}

#[pure]
fn max(a: i32, b: i32) -> i32 {
    if a > b {
        a
    } else {
        b
    }
}

fn test_max1() {
    let a = 5;
    let b = 6;
    let z = max(a, b);
    assert!(z == 6);
}

fn test_max2() {
    let a = 5;
    let b = 6;
    let z = max(a, b);
    assert!(z == 5);
}

#[ensures(

    true && (true ==>

        result == 3     // test that we get correct span information
            )

    && (true || false)

        )

    ]
fn test_max3() -> i32 {
    let a = 4;
    let b = 3;
    max(a, b)
}

#[requires(a > b)]
#[ensures(result == max(a, b))]
fn test_max4(a: i32, b: i32) -> i32 {
    a
}

#[requires(a < b)]
#[ensures(result == max(a, b))]
fn test_max5(a: i32, b: i32) -> i32 {
    a
}

fn main() {}
