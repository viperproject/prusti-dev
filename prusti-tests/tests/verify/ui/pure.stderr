error: [Prusti: verification error] the asserted expression might not hold
  --> $DIR/pure.rs:39:5
   |
39 |     assert!(z == 5);
   |     ^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/pure.rs:46:9
   |
46 |         result == 3     // test that we get correct span information
   |         ^^^^^^^^^^^
   |
note: the error originates here
  --> $DIR/pure.rs:54:1
   |
54 | / fn test_max3() -> i32 {
55 | |     let a = 4;
56 | |     let b = 3;
57 | |     max(a, b)
58 | | }
   | |_^

error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/pure.rs:67:11
   |
67 | #[ensures(result == max(a, b))]
   |           ^^^^^^^^^^^^^^^^^^^
   |
note: the error originates here
  --> $DIR/pure.rs:68:1
   |
68 | / fn test_max5(a: i32, b: i32) -> i32 {
69 | |     a
70 | | }
   | |_^

error: aborting due to 3 previous errors

